blockchain/blockchain.go

```diff
@@ -1763,38 +1756,40 @@ func (bc *BlockChain) CloseBlockSubscriptionLoop() {
 }
 
 // replaceCurrentBlock replaces bc.currentBlock to the given block.
-// func (bc *BlockChain) replaceCurrentBlock(latestBlock *types.Block) {
-// 	bc.mu.Lock()
-// 	defer bc.mu.Unlock()
-//
-// 	if latestBlock == nil {
-// 		logger.Error("no latest block")
-// 		return
-// 	}
-//
-// 	// Return early if it is the first block update.
-// 	currentBlock := bc.CurrentBlock()
-// 	if currentBlock == nil {
-// 		bc.insert(latestBlock)
-// 		logger.Info("inserted an initial block", "blockNumber", latestBlock.NumberU64(),
-// 			"hash", latestBlock.Hash().String(), "stateRoot", latestBlock.Root().String())
-// 		return
-// 	}
-//
-// 	// Don't update current block if the latest block is not newer than current block.
-// 	if currentBlock.NumberU64() >= latestBlock.NumberU64() {
-// 		logger.Debug("")
-// 		return
-// 	}
-//
-// 	// Insert a new block and update metrics
-// 	bc.insert(latestBlock)
-// 	headBlockNumberGauge.Update(latestBlock.Number().Int64())
-// 	bc.stateCache.TrieDB().UpdateMetricNodes()
-//
-// 	logger.Info("inserted a new block", "prevBlockNumber", currentBlock.NumberU64(), "blockNumber",
-// 		latestBlock.NumberU64(), "hash", latestBlock.Hash().String(), "stateRoot", latestBlock.Root().String())
-// }
+func (bc *BlockChain) replaceCurrentBlock(latestBlock *types.Block) {
+	bc.mu.Lock()
+	defer bc.mu.Unlock()
+
+	if latestBlock == nil {
+		logger.Error("no latest block")
+		return
+	}
+
+	// Return early if it is the first block update.
+	currentBlock := bc.CurrentBlock()
+	if currentBlock == nil {
```

Comment:
When is this current block nill?
It is the moment before `loadLastState`?

In that case, how about just returning?

It seems that `loadLastState()` and this function has a race condition of `currentBlock`.
Is it ok?

Reply:
I removed the logic. It was for an edge case, but the case is not realistic now. 
This code is executed after `loadLastState()` and there is no race condition.